File: roses.js
if(!Array.prototype.last) {
    Array.prototype.last = function() {
        return this[this.length - 1];
    }
}

function caps(a) {return a.substring(0,1).toUpperCase() + a.substring(1,a.length);}
function uniform(a, b) { return ( (Math.random()*(b-a))+a ); }
function showSlide(id) { $(".slide").hide(); $("#"+id).show(); }
function shuffle(v) { newarray = v.slice(0);for(var j, x, i = newarray.length; i; j = parseInt(Math.random() * i), x = newarray[--i], newarray[i] = newarray[j], newarray[j] = x);return newarray;} // non-destructive.
function sample(v) {return(shuffle(v)[0]);}
function rm(v, item) {if (v.indexOf(item) > -1) { v.splice(v.indexOf(item), 1); }}
function rm_sample(v) {var item = sample(v); rm(v, item); return item;}
var startTime;

var Target = function(frequent, infrequent, superset, beginning, end, article) {
  this.frequent = frequent;
  this.infrequent = infrequent;
  this.superset = superset;
  this.beginning = beginning;
  this.end = end;
  this.article = article == null ? function() {return "";} : article;

  this.sentence = function(version) {
    var frequency = version.frequency; //frequent or infrequent
    var hasOther = version.hasOther; //true or false
    var order = version.order; //subSuper or superSub
    var conjunction = hasOther ? " and other " : " and ";
    var a;
    var b;
    if (order == "subSuper") {
      b = this.superset;
      a = frequency == "frequent" ? this.frequent : this.infrequent;
    } else {
      a = this.superset;
      b = frequency == "frequent" ? this.frequent : this.infrequent;
    }
    if (this.beginning == "") {
      a = caps(a);
    }
    return this.beginning + this.article(a) + a + conjunction + b + this.end;
  }
}

var targetTypes = [];
for (var i=0; i<2; i++) {
  for (var j=0; j<2; j++) {
    for (var k=0; k<2; k++) {
      var frequency = ["frequent","infrequent"][i];
      var hasOther = [true, false][j];
      var order = ["subSuper", "superSub"][k];
      if (!(order == "superSub" && hasOther == true)) {
        targetTypes.push({
          frequency:frequency,
          hasOther:hasOther,
          order:order
        })
      }
    }
  }
}
var qTypes = shuffle(targetTypes.concat(["filler0", "filler1", "filler2", "filler3", "filler4", "filler5"]));
var fillers = {"filler0": "Pat telephoned the sister of Sally’s friend yesterday."
  , "filler1": "The gopher dug a tunnel underneath the fence."
  , "filler2": "A math textbook and a spiral notebook were lying on the kitchen table."
  , "filler3": "Adam and Charlie liked to play cops and robbers together when they were little."
  , "filler4": "Nobody knew the solution to any of the logic puzzles in the book."
  , "filler5": "Jill and Tom met at a cafe for their first date."}
var targets = shuffle([ new Target("roses", "daffodils", "flowers", "This shop sells ", ".")
  , new Target("biologists", "paleontologists", "scientists", "The advisory panel included a number of ", ".")
  , new Target("beef", "veal", "meat", "The recipe called for ", " to be included in the stew.")
  , new Target("oaks", "spruces", "trees", "", " lined the path through the forest.")
  , new Target("surgeons", "anesthesiologists", "doctors", "At the conference there was a special informational session for ", ".")
  , new Target("horse", "waterfowl", "animal", "Lilly ran ", " rescue operation in the small town.", function(nextWord) {
    if (nextWord == "animal") {
      return "an ";
    } else {
      return "a "
    }
  })]);

nQs = 12;

$(document).ready(function() {
  showSlide("consent");
  $("#mustaccept").hide();
  startTime = Date.now();
});

var experiment = {
  data: {
    trials:[]
  },
  
  instructions: function() {
    if (turk.previewMode) {
      $("#instructions #mustaccept").show();
    } else {
      showSlide("instructions");
      $("#begin").click(function() { experiment.trial(0); })
    }
  },
  
  trial: function(qNumber) {
    $('.bar').css('width', ( (qNumber / nQs)*100 + "%"));
    $(".err").hide();

    var trialType = qTypes.shift();
    var trialStart = Date.now();
    var trialData = {response:[], rt:[], qNumber:qNumber, trialType:trialType};

    $("#trialInstructions").html("How acceptable do you think the following sentence is?");

    var sentence;
    console.log(trialType);
    if (typeof(trialType) == "string") {
      sentence = fillers[trialType];
    } else {
      var target = targets.shift();
      console.log(qNumber);
      console.log(trialType);
      console.log(target);
      sentence = target.sentence(trialType);
    }
    $("#sentence").html(sentence);

    showSlide("trial");
    var responseNeeded = true;

    $("#sliderContainer").html('<div id="responseSlider"></div>');
    $("#responseSlider").slider({
               animate: true,
               max: 1 , min: 0, step: .01, value: 0.5,
               slide: function( event, ui ) {
                   $("#responseSlider .ui-slider-handle").css({
                      "background":"#E0F5FF",
                      "border-color": "#001F29"
                   });
               },
               change: function( event, ui ) {
                   $("#responseSlider").css({"background":"#99D6EB"});
                   $("#responseSlider .ui-slider-handle").css({
                     "background":"#667D94",
                     "border-color": "#001F29" });
                   responseNeeded = false;
                   trialData["response"].push(ui.value);
                   trialData["rt"].push(Date.now() - trialStart);
               }});

    $(".continue").click(function() {
      if (responseNeeded) {
        $(".err").show();
      } else {
        $(".continue").unbind("click");
        $(".err").hide();
        experiment.data["trials"].push(trialData);
        if (qNumber + 1 < nQs) {
          experiment.trial(qNumber+1);
        } else {
          experiment.questionaire();
        }
      }
    })
  },
  
  questionaire: function() {
    //disable return key
    $(document).keypress( function(event){
     if (event.which == '13') {
        event.preventDefault();
      }
    });
    //progress bar complete
    $('.bar').css('width', ( "100%"));
    showSlide("questionaire");
    $("#formsubmit").click(function() {
      rawResponse = $("#questionaireform").serialize();
      pieces = rawResponse.split("&");
      var age = pieces[0].split("=")[1];
      var lang = pieces[1].split("=")[1];
      var comments = pieces[2].split("=")[1];
      if (lang.length > 0) {
        experiment.data["language"] = lang;
        experiment.data["comments"] = comments;
        experiment.data["age"] = age;
        var endTime = Date.now();
        experiment.data["duration"] = endTime - startTime;
        showSlide("finished");
        setTimeout(function() { turk.submit(experiment.data) }, 1000);
      }
    });
  }
}
  

File: roses.R
library('lme4')
library('ggplot2')

item = "flowers"
freq = "roses"
infreq = "daffodils"

df = read.table("roses-4-14.results", sep="\t", header=T)
#df = df[as.numeric(as.character(df$subject))>39,]
r = df[is.na(df$filler) & !is.na(df$frequency),]
#r = r[r$item == item,]
#r = r[r$item != "animal",]
filler = df[!(is.na(df$filler)),]
#print(summary(r))

# fit = lm(rating ~ frequency * order * hasOther, data=r)
# print(anova(fit))

conf <- function(v) {
  v <- v[is.na(v) == F]
  nsubj=length(v)
  sample.means <- replicate(100, mean(sample(v, nsubj, replace=TRUE)))
  return(quantile(sample.means, c(0.025, 0.975)))
}
lower.conf <- function(v) {
  conf(v)[["2.5%"]]
}
upper.conf <- function(v) {
  conf(v)[["97.5%"]]
}

pretty.name = function(f, ord, other) {
  if (f == "frequent") {
    if (ord == "subSuper") {
      if (other == "True") {
        return(paste(freq, "and other", item))
      } else {
        return(paste(freq, "and", item))
      }
    } else {
      return(paste(item, "and", freq))
    }
  } else {
    if (ord == "subSuper") {
      if (other == "True") {
        return(paste(infreq, "and other", item))
      } else {
        return(paste(infreq, "and", item))
      }
    } else {
      return(paste(item, "and", infreq))
    }
  }
}
r$cond = sapply(1:nrow(r), function(i) {
  pretty.name(r$frequency[i], r$order[i], r$hasOther[i])
})
r$cond = factor(r$cond, levels=c(
  paste(freq, "and", item),
  paste(freq, "and other", item),
  paste(infreq, "and", item),
  paste(infreq, "and other", item),
  paste(item, "and", freq),
  paste(item, "and", infreq)
#   "roses and flowers",
#   "roses and other flowers",
#   "daffodils and flowers",
#   "daffodils and other flowers",
#   "flowers and roses",
#   "flowers and daffodils"
))

filler.map = list(
  "adam and charlie"="Adam and Charlie liked to play cops and robbers together when they were little.",
  "jill and tom"="Jill and Tom met at a cafe for their first date.",
  "math textbook"="A math textbook and a spiral notebook were lying on the kitchen table.",
  "nobody knew"="Nobody knew the solution to any of the logic puzzles in the book.",
  "pat telephoned"="Pat telephoned the sister of Sally's friend yesterday.",
  "the gopher"="The gopher dug a tunnel underneath the fence."
)

#dependent measure = rating
#filler$filler = factor(filler$filler, levels=levels(filler$filler), label=sapply(levels(filler$filler), function(f) {return(filler.map[[f]])}))
filler$cond = filler$filler
target.rating.means = aggregate(rating ~ cond, data=r, mean)
target.rating.upper = aggregate(rating ~ cond, data=r, upper.conf)
target.rating.lower = aggregate(rating ~ cond, data=r, lower.conf)
filler.rating.means = aggregate(rating ~ cond, data=filler, mean)
filler.rating.upper = aggregate(rating ~ cond, data=filler, upper.conf)
filler.rating.lower = aggregate(rating ~ cond, data=filler, lower.conf)
rating.means = rbind(target.rating.means, filler.rating.means)
rating.upper = rbind(target.rating.upper, filler.rating.upper)
rating.lower = rbind(target.rating.lower, filler.rating.lower)

ggplot(data=target.rating.means, aes(x=cond, y=rating, fill=cond)) +
  geom_bar(colour="black", stat="identity") +
  ggtitle("aggregated ratings") +
  #ggtitle(paste("ratings for ",item)) +
  theme(plot.title = element_text(face="bold")) +
  #scale_fill_discrete(name="Trial Type", labels=c(sapply(levels(filler$cond), function(f) {return(paste("FILLER:", filler.map[[f]]))}))) +
  scale_fill_manual(values=c("#fb9a99",
                             "#e31a1c",
                             "#b2df8a", "#33a02c",
                             "#a6cee3",
                             "#1f78b4")) + 
  geom_errorbar(aes(ymin=target.rating.lower$rating, ymax=target.rating.upper$rating), width=.1) +
  geom_line() +
  geom_point()

ggplot.rainbow = function(n) {
  return(sapply(0:(n-1), function(i) {
#     return(hcl(h=(i*(360/n)), c=100, l=65))
    return(hcl(h=((360-(i+1))*(360/n)), c=100, l=65))
  }))
}

ggplot(data=rating.means, aes(x=cond, y=rating, fill=cond)) +
  geom_bar(colour="black", stat="identity") +
  #ggtitle(paste("ratings for ",item)) +
  ggtitle("aggregated ratings with fillers") +
  #theme(plot.title = element_text(face="bold")) +
#   scale_fill_discrete(
#     name="Trial Type",
#     labels=c(levels(r$cond), sapply(levels(filler$cond), function(f) {return(paste("FILLER:", filler.map[[f]]))}))
#   ) +
  scale_fill_manual(
    name="Trial Type",
    labels=c(levels(r$cond), sapply(levels(filler$cond), function(f) {return(paste("FILLER:", filler.map[[f]]))})),
    values=c(c("#fb9a99", "#e31a1c", "#b2df8a", "#33a02c", "#a6cee3", "#1f78b4"),
             ggplot.rainbow(6))) + 
  geom_errorbar(aes(ymin=rating.lower$rating, ymax=rating.upper$rating), width=.1) +
  geom_line() +
  geom_point()

#dependent measure = rt
target.rt.means = aggregate(rt ~ cond, data=r, mean)
target.rt.upper = aggregate(rt ~ cond, data=r, upper.conf)
target.rt.lower = aggregate(rt ~ cond, data=r, lower.conf)
filler.rt.means = aggregate(rt ~ cond, data=filler, mean)
filler.rt.upper = aggregate(rt ~ cond, data=filler, upper.conf)
filler.rt.lower = aggregate(rt ~ cond, data=filler, lower.conf)
rt.means = rbind(target.rt.means, filler.rt.means)
rt.upper = rbind(target.rt.upper, filler.rt.upper)
rt.lower = rbind(target.rt.lower, filler.rt.lower)

ggplot(data=rt.means, aes(x=cond, y=rt, fill=cond)) +
  geom_bar(colour="black", stat="identity") +
  ggtitle("Reaction Time") +
  theme(plot.title = element_text(face="bold")) +
  scale_fill_manual(values=c("#fb9a99", "#e31a1c", "#b2df8a", "#33a02c", "#a6cee3", "#1f78b4")) + 
  geom_errorbar(aes(ymin=rt.lower$rt, ymax=rt.upper$rt), width=.1) +
  geom_line() +
  geom_point()

ggplot(data=rt.means, aes(x=cond, y=rt, fill=cond)) +
  geom_bar(colour="black", stat="identity") +
  ggtitle("Reaction Time") +
  theme(plot.title = element_text(face="bold")) +
  scale_fill_manual(
    name="Trial Type",
    labels=c(levels(r$cond), sapply(levels(filler$cond), function(f) {return(paste("FILLER:", filler.map[[f]]))})),
    values=c(c("#fb9a99", "#e31a1c", "#b2df8a", "#33a02c", "#a6cee3", "#1f78b4"),
             ggplot.rainbow(6))) + 
  geom_errorbar(aes(ymin=rt.lower$rt, ymax=rt.upper$rt), width=.1) +
  geom_line() +
  geom_point()
